db/cache_db.go

```diff
@@ -0,0 +1,172 @@
+package db
+
+import (
+	"fmt"
+	"sort"
+	"sync"
+)
+
+type cDBValue struct {
+	value   []byte
+	deleted bool
+}
+
+// CacheDB wraps an in-memory cache around an underlying database.
+// NOTE: Needed by MultiStore.CacheWrap().
+type CacheDB struct {
+	mtx    sync.Mutex
+	cache  map[string]cDBValue
+	parent DB
+}
+
+// NewCacheDB returns a new in-memory cache.
+func NewCacheDB(parent DB) *CacheDB {
+	db := &CacheDB{
+		cache:  make(map[string]cDBValue),
+		parent: parent,
+	}
+	return db
+}
+
+// Get gets the value for the given key. It retrieves the value from the cache or accesses the
+// underlying database.
+func (db *CacheDB) Get(key []byte) []byte {
+	db.mtx.Lock()
+	defer db.mtx.Unlock()
+
+	value, ok := db.cache[string(key)]
+	if ok {
+		return value.value
+	}
+	return db.parent.Get(key)
```

Comment:
Hmm, I don't see where we memoize the cache-missed value for this, perhaps let's do this.
```go
value, ok := db.cache[string(key)]
if !ok {
   data := db.parent.Get(key)
   value = cDBValue{value: data}
   db.cache[string(key)] = value
}
return value.value
```

Reply:
Also should we perhaps check if `deleted { return nil }`?

Reply:
I was wondering about that too. Maybe @jaekwon can clarify how this functionality is supposed to work. I got the understanding that you had to explicitly add values to the cache.

Reply:
Doh, yeah.

Reply:
if deleted, value.value will be nil.